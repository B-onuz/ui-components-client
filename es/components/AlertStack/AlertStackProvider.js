"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _AlertStackContext = _interopRequireDefault(require("./AlertStackContext"));

var _uuid = require("uuid");

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/AlertStack/AlertStackProvider.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var AlertStackProvider = function AlertStackProvider(_ref) {
  var children = _ref.children;

  var _useState = (0, _react.useState)([]),
      topRightMessages = _useState[0],
      setTopRightMessages = _useState[1];

  var _useState2 = (0, _react.useState)([]),
      topLeftMessages = _useState2[0],
      setTopLeftMessages = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      bottomRightMessages = _useState3[0],
      setBottomRightMessages = _useState3[1];

  var _useState4 = (0, _react.useState)([]),
      bottomLeftMessages = _useState4[0],
      setBottomLeftMessages = _useState4[1];

  var _hideMessage = function _hideMessage(setMessages, _id) {
    setMessages(function (messages) {
      return messages.filter(function (message) {
        return message._id !== _id;
      });
    });
  };

  var _showMessage = function _showMessage(fn, params) {
    fn(function (messages) {
      return [].concat(messages, [params]);
    });

    if (params.time) {
      setTimeout(function () {
        _hideMessage(fn, params._id);
      }, params.time || params.time);
    }
  };

  var showMessage = function showMessage(_ref2) {
    var _ref2$position = _ref2.position,
        position = _ref2$position === void 0 ? 'bottom-left' : _ref2$position,
        title = _ref2.title,
        message = _ref2.message,
        color = _ref2.color,
        time = _ref2.time;
    var messageParam = {
      _id: (0, _uuid.v4)(),
      title: title,
      message: message,
      color: color,
      time: time
    };

    switch (position) {
      case 'top-right':
        return _showMessage(setTopRightMessages, messageParam);

      case 'top-left':
        return _showMessage(setTopLeftMessages, messageParam);

      case 'bottom-right':
        return _showMessage(setBottomRightMessages, messageParam);

      case 'bottom-left':
        return _showMessage(setBottomLeftMessages, messageParam);

      default:
        return _showMessage(setBottomLeftMessages, messageParam);
    }
  };

  return /*#__PURE__*/_react["default"].createElement(_AlertStackContext["default"].Provider, {
    value: {
      showMessage: showMessage,
      topRightMessages: topRightMessages,
      topLeftMessages: topLeftMessages,
      bottomRightMessages: bottomRightMessages,
      bottomLeftMessages: bottomLeftMessages
    },
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 44,
      columnNumber: 5
    }
  }, children);
};

var _default = AlertStackProvider;
exports["default"] = _default;