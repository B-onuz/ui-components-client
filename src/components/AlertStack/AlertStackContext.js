import React from 'react'

const AlertStackContext = React.createContext()

export default AlertStackContext
