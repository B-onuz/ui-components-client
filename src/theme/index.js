const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
