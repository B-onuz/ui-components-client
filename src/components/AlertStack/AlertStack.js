import React, { useState, useEffect, useContext } from 'react'
import styled from 'styled-components'
import { color } from 'styled-system'
import ReactDOM from 'react-dom'
import Alert from '../Alert'
import AlertStackContext from './AlertStackContext'

const Wrapper = styled.div`
  position: fixed;
  z-index: 9999999;
  max-width: 100%;
  margin: 10px;
  ${({ top, right, bottom, left }) => {
    if (top && right) {
      return `top: ${0}px; right: ${0}px;`
    }
    if (left && top) {
      return `left: ${0}px; top: ${0}px;`
    }
    if (bottom && left) {
      return `bottom: ${0}px; left: ${0}px;`
    }
    if (bottom && right) {
      return `bottom: ${0}px; right: ${0}px;`
    }
    return `bottom: ${0}px; left: ${0}px;`
  }}
`

const useRootAlertStack = () => {
  const [$rootAlertStack, setRootAlertStack] = useState()

  useEffect(() => {
    const rootAlertStack = document.getElementById('root-alert-stack')
    if (rootAlertStack) {
      setRootAlertStack(rootAlertStack)
    } else {
      const rootEl = document.createElement('div')
      rootEl.id = 'root-alert-stack'
      setRootAlertStack(rootEl)
      document.body.appendChild(rootEl)
    }
  })

  return { $rootAlertStack }
}

const AlertStack = ({ ...rest }) => {
  const { $rootAlertStack } = useRootAlertStack()
  const { topRightMessages, topLeftMessages, bottomLeftMessages, bottomRightMessages } = useContext(AlertStackContext)

  if (!$rootAlertStack) return null
  return ReactDOM.createPortal(
    <>
      {!!topRightMessages.length && (
        <Wrapper top right>
          {topRightMessages.map(({ title, message, color }) => (
            <Alert mt={10} animate={'top'} color={color} title={title} message={message} />
          ))}
        </Wrapper>
      )}
      {!!topLeftMessages.length && (
        <Wrapper top left>
          {topLeftMessages.map(({ title, message, color }) => (
            <Alert mt={10} animate={'top'} color={color} title={title} message={message} />
          ))}
        </Wrapper>
      )}
      {!!bottomLeftMessages.length && (
        <Wrapper bottom left>
          {bottomLeftMessages.map(({ title, message, color }) => (
            <Alert mt={10} animate={'bottom'} color={color} title={title} message={message} />
          ))}
        </Wrapper>
      )}
      {!!bottomRightMessages.length && (
        <Wrapper bottom right>
          {bottomRightMessages.map(({ title, message, color }) => (
            <Alert mt={10} animate={'bottom'} color={color} title={title} message={message} />
          ))}
        </Wrapper>
      )}
    </>,
    $rootAlertStack
  )
}

export default AlertStack
