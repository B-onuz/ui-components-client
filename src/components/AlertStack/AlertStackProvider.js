import React, { useState, useEffect } from 'react'
import AlertStackContext from './AlertStackContext'
import { v4 as uuidv4 } from 'uuid'

const AlertStackProvider = ({ children }) => {
  const [topRightMessages, setTopRightMessages] = useState([])
  const [topLeftMessages, setTopLeftMessages] = useState([])
  const [bottomRightMessages, setBottomRightMessages] = useState([])
  const [bottomLeftMessages, setBottomLeftMessages] = useState([])

  const _hideMessage = (setMessages, _id) => {
    setMessages((messages) => messages.filter((message) => message._id !== _id))
  }

  const _showMessage = (fn, params) => {
    fn((messages) => {
      return [...messages, params]
    })

    if (params.time) {
      setTimeout(() => {
        _hideMessage(fn, params._id)
      }, params.time || params.time)
    }
  }

  const showMessage = ({ position = 'bottom-left', title, message, color, time }) => {
    const messageParam = { _id: uuidv4(), title, message, color, time }
    switch (position) {
      case 'top-right':
        return _showMessage(setTopRightMessages, messageParam)
      case 'top-left':
        return _showMessage(setTopLeftMessages, messageParam)
      case 'bottom-right':
        return _showMessage(setBottomRightMessages, messageParam)
      case 'bottom-left':
        return _showMessage(setBottomLeftMessages, messageParam)
      default:
        return _showMessage(setBottomLeftMessages, messageParam)
    }
  }

  return (
    <AlertStackContext.Provider value={{ showMessage, topRightMessages, topLeftMessages, bottomRightMessages, bottomLeftMessages }}>
      {children}
    </AlertStackContext.Provider>
  )
}

export default AlertStackProvider
