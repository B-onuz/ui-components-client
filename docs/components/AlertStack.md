# AlertStack (Component)

Controle de mensagens de erro, sucesso, avisos etc.

## Configuração

```jsx
/* ... */
import React from 'react'
import { ThemeProvider } from 'styled-components'
import 'jest-styled-components'
import theme from 'ui-components/lib/theme'
import AlertStackProvider from 'ui-components/lib/components/AlertStack/AlertStackProvider'
import AlertStack from 'ui-components/lib/components/AlertStack'

ReactDOM.render(
  <React.StrictMode>
    <ThemeProvider theme={theme}>
      <AlertStackProvider>
        <App />
        <AlertStack />
      </AlertStackProvider>
    </ThemeProvider>
  </React.StrictMode>,
  document.getElementById('root')
)

/* ... */
```

```jsx
import React from 'react'
import useAlertStack from 'ui-components/lib/hooks/useAlertStack'

// este componente deve estar dentro da hierarquia de componentes.

const Screen = () => {
  const { showMessage } = useAlertStack()
  return (
    <Button
      onClick={() =>
        showMessage({
          title: 'Cadastro realizado com sucesso', //  Título da mensagem
          message: 'Mensagem mais detalhada aqui', // Mensagem detalhada
          color: 'success', // color, veja theme.color
          position: 'bottom-right', // posição sempre nessa ordem top ou bottom seguido de right ou left
          time: 3000, // tempo em ms que a mensagem permanecerá exibida.
        })
      }
    >
      Mostrar Mensagem
    </Button>
  )
}
```
