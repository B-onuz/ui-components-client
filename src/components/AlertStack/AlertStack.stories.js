import React, { useState } from 'react'
import Button from '../Button'
import TextField from '../TextField'
import SelectField from '../SelectField'
import AlertStackProvider from './AlertStackProvider'
import AlertStack from './AlertStack'
import useAlertStack from '../../hooks/useAlertStack'

export default {
  title: 'AlertStack',
  component: AlertStack,
}

const App = () => {
  const { showMessage } = useAlertStack()
  const [message, setMessage] = useState('Lorem ipsum dolor sit amet larem')
  const [title, setTitle] = useState('Lorem ipsum dolor')
  const [color, setColor] = useState('success')
  const [position, setPosition] = useState('bottom-left')
  const [time, setTime] = useState(3000)
  return (
    <>
      <TextField m={20} label="Mensagem" value={message} onChange={(e) => setMessage(e.target.value)} />
      <TextField m={20} label="Título" value={title} onChange={(e) => setTitle(e.target.value)} />
      <TextField m={20} label="Duração" type={'number'} value={time} onChange={(e) => setTime(e.target.value)} />
      <SelectField m={20} value={color} onChange={(e) => setColor(e.target.value)} required label={'Color'} m={2} placeholder={'Nome'}>
        <option value="danger">Danger</option>
        <option value="warning">Warning</option>
        <option value="success">Success</option>
        <option value="info">Info</option>
      </SelectField>
      <SelectField m={20} value={position} onChange={(e) => setPosition(e.target.value)} required label={'Posição'} m={2} placeholder={'Nome'}>
        <option value="top-right">Top - Right</option>
        <option value="top-left">Top - Left</option>
        <option value="bottom-left">Bottom - Left</option>
        <option value="bottom-right">Bottom - Right</option>
      </SelectField>
      <Button
        onClick={() =>
          showMessage({
            title,
            message,
            color,
            position,
            time: Number(time),
          })
        }
      >
        Mostrar Mensagem
      </Button>
    </>
  )
}

export const AlertStackExample = () => {
  return (
    <AlertStackProvider>
      <App />
      <AlertStack />
    </AlertStackProvider>
  )
}
