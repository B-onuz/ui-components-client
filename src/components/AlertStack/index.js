import AlertStack from './AlertStack'
import AlertStackContext from './AlertStackContext'

export { AlertStackContext }

export default AlertStack
