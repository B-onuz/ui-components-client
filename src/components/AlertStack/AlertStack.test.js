import React from 'react'
import AlertStack from './AlertStack.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'
import AlertStackProvider from './AlertStackProvider.js'

test('Should render AlertStack', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <AlertStackProvider>
        <AlertStack />
      </AlertStackProvider>
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
