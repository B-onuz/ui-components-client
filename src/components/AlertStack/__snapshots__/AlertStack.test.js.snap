// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AlertStack 1`] = `null`;
